content/Stop creating services for everything - Knowing YAAF & the FormObject Pattern.md

```diff
@@ -0,0 +1,143 @@
+[![YAAF](images/Yaaf-Blog-cover.png)](https://github.com/rootstrap/yaaf)
+
+# Form object patterns in Rails: Stop creating services for everything with YAAF
+
+## Working with large controllers
+
+How many times have you encountered large controller methods? If you are _lucky_ like me, probably many times.
+
+One of the most common practices to start refactoring a long controller is to move the code to a service.
+
+Services are great, and if we code them in an atomic way, they will be easy to test and understand. But the problem is when we use the services it's like using a _Swiss Army knife_.
+
+>-"Hey, I don't know how to properly refactor this piece of code"
+>-"Dude, just do a new service"
+
+**But no!**, making a new service is not always the best option. In some cases, we are _reinventing the wheel_ and maybe there is a pattern that already fits with our needs.
+
+So here comes **YAAF** (Yet Another Active Form) to save our day. **YAAF** is a gem that lets you create form objects in an easy and friendly Rails way. It makes use of `ActiveRecord` and `ActiveModel` features in order to provide you with a form object that behaves pretty much like a Rails model and still be completely configurable.
+
+## When to use YAAF?
+
+Let's imagine that we have an API endpoint that saves a new post on our database. A post has a title, body, publisher, and could also have tags and a category.
+
+Tags and categories could be created at the moment that the publisher sends the post. (If our inputs don't find the correct tag or category, they will let the user write the name of a new one).
+
+So in our controller, in the worst-case scenario, we could have something like this:
+
+```ruby
+class Api::V1::PostsController < Api::V1::ApiController
+  def index
+    @posts = Post.all
+  end
+
+  def create
+    ActiveRecord::Base.transaction do
+      @post = Post.new(post_params)
+      @post.tags = params[:tags].map do |tag|
+        tag[:id].present? ? Tag.find(tag[:id]) : Tag.find_or_create_by(name: tag[:name])
+      end
+      if params[:category_name].present?
+        category = Category.create!(name: params[:category_name])
+        @post.category = category
+      end
+      @post.save!
+    end
+  end
+
+
+  def post_params
+    params.require(:post).permit(:title, :body, :publisher_id, :category_id)
+  end
+end
+```
+
+It looks terrible, doesn't it? Maybe our first thinking about that piece of code is to make a refactoring moving the creation of the post to a service named `PostCreationService`, which could be useful and it might be used in the future in another part of the system. But... what have we said about _reinventing the wheel_?
```

Comment:
```suggestion
It looks terrible, right? Maybe our first thinking about that piece of code is to make a refactoring which moves the creation of the post to a service named `PostCreationService`. This could be useful and might be used in the future in another part of the system. But... what have we said about _reinventing the wheel_?
```